### ISSUE TITLE
Reduced CI workflow for documentation PRs

### ISSUE THREAD

**Issue by cameel** [ethereum/solidity#12424]
Currently we run all CI jobs on all PRs. CircleCI has recently introduced the [dynamic configuration](https://circleci.com/docs/2.0/dynamic-config/) feature which allows us to select different sets of CI jobs depending on various criteria like e.g. modified paths. We can use this to make CI finish faster (and use less credits) on documentation PRs.

Add a new workflow to [`.circleci/config.yml`](https://github.com/ethereum/solidity/blob/develop/.circleci/config.yml) called `docs`. The workflow should be triggered when changes are made only to files inside `docs/` and/or text files at the root of the repository. The workflow should include only the following jobs:
- `b_docs`
- `chk_spelling`
- `chk_docs_examples`
- `chk_coding_style`
- `chk_pylint`
- `chk_antlr_grammar`
- `chk_buglist`
- `chk_docs_pragma_min_version`
- `t_ubu_pyscripts`
- `t_win_pyscripts`

The last two are not strictly necessary right now but they're are very light and we do have some python stuff inside `docs/` and we might start testing at some point in the future.

This approach has one downside - we won't be running command-line tests and a part of that is a check that all the examples in the docs do compile using the current build of the compiler. I think that his is a reasonable trade-off and any problems in the docs that would fail this check will soon be discovered anyway during the next full run. If we don't want to make that trade-off we can also make the new workflow include a build for one platform and a limited run of command-line tests.

We can limit the impact of the above by always running the main workflow on `develop`. So even if problems are not detected in the PR, they'll be caught immediately after it's merged.

**Comment by cameel** [comment]
If this works out, we could introduce reduced workflows for a few more cases. Some ideas:
- When only test cases are modified, we can skip some checks like external tests, bytecode, docs, etc.
- If only .py files are modified we could limit check to just pylint and running Python unit tests.
- If only grammar is modified, we could only run `b_docs` and the ANTLR job.
- If only buglist is modified we could run only its job.

**Comment by chriseth** [comment]
Note that `chk_docs_pragma_min_version` needs a working solidity binary.
Don't we also have a test that checks that all documentation snippets compile without error?

Note that PRs against develop should use the develop binary and PRs against breaking should use the breaking binary. Is it possible to take the binary from the last CircleCI run of the target branch or something?

Also this restriction should be disabled for develop - we should always have a full run on develop.

**Comment by cameel** [comment]
> Note that `chk_docs_pragma_min_version` needs a working solidity binary.

Does it actually? Looking at the script, it can use local `solc-$version` if that's just lying around and `$version` can be the dev version returned by `get_version.sh` but why would the binary be named like that? It looks to me like it'll just always try to download from github: https://github.com/ethereum/solidity/blob/018f8cd2e20b577f8e13759a7bd88b1d9959f8e2/scripts/docs_version_pragma_check.sh#L178-L185

> Don't we also have a test that checks that all documentation snippets compile without error?

We do and the fact that we do that separately is what makes me think that the above does not use local build of the compiler.
https://github.com/ethereum/solidity/blob/018f8cd2e20b577f8e13759a7bd88b1d9959f8e2/test/cmdlineTests.sh#L489-L530

> Note that PRs against develop should use the develop binary and PRs against breaking should use the breaking binary. Is it possible to take the binary from the last CircleCI run of the target branch or something?

It is but it would require downloading artifacts via CircleCI API. Coincidentally, I have a script just for that :P

> Also this restriction should be disabled for develop - we should always have a full run on develop.

Good point. This should only be done on PRs. I'll add that to the description.

**Comment by cameel** [comment]
We just talked about this with @Marenz in the context of adding a reduced workflow for LSP PRs.

> Note that PRs against develop should use the develop binary and PRs against breaking should use the breaking binary. Is it possible to take the binary from the last CircleCI run of the target branch or something?

It requires accessing CircleCI API but it's doable. And given that we have recently merged [`download_benchmarks.py`](https://github.com/ethereum/solidity/blob/develop/scripts/externalTests/download_benchmarks.py), we already have the base code ready for doing that. Now it's enough to slightly adjust the script to handle arbitrary artifacts.

> Note that `chk_docs_pragma_min_version` needs a working solidity binary.

Also, an important point is that we need *a* binary. Not all of them. Even if we built it from scratch, skipping win/mac/wasm builds, bytecode comparison and external tests is still a significant saving.
